doc/port_link_training/port-link-training-design.md

```diff
@@ -0,0 +1,369 @@
+
+On the other hand, a new port attribute will be introduced for the link-training ability query, and please note the link-training ability could be unavailable on the certain ports on a switch silicon (e.g. The uplink/management ports), hence we need to do the query at per-port basis.  
+
+```cpp
+    /**
+     * @brief Query link-training support
+     *
+     * @type bool
+     * @flags READ_ONLY
+     */
+    SAI_PORT_ATTR_SUPPORTED_LINK_TRAINING_MODE,
+```
+
+Vendor-specific SAI implementation is not in the scope of this document, but there are some common requirements for SAI:
+
+1. swss must check the link-training abilities before making the corresponding request to syncd.
+2. SAI implementation must return error code if any of the above attributes is not supported, swss and syncd must not crash.
+3. SAI implementation must keep port link training disabled by default for backward compatibility. As long as swss and SAI keep backward compatible, user need not change anything after this feature is implemented and available in SONiC.
+
+### Configuration and management 
+
+#### CLI Enhancements
+
+A few new CLI commands are designed to support port link training.
+
+##### Config link training mode
+
+
+```
+Format:
+  config interface link-training <interface_name> <mode> [-f]
+
+Arguments:
+  interface_name: [mandatory] name of the interface to be configured. e.g: Ethernet0
+  mode: [mandatory] link training mode, can be either "on" or "off"
+  -f: [optional] forcing link-training configuration on an unsupported transceiver
+
+Example:
+  config interface link-training Ethernet0 on
+  config interface link-training Ethernet0 on -f
+  config interface link-training Ethernet0 off
+
+Return:
+  - error message if interface_name or mode is invalid
+  - error message if the link-training is not supported on the interface
+  - empty upon success
+```
+
+For deployment considerations, when the link-training is enabled on a transceiver without LT capabilities, this config command should report an error and get aborted, unless '-f' option is specified.
+
+##### Show link training status
+
+A new CLI command will be added to display the port link training status. All data of this command are fetched from **STATE_DB**.
+
+```
+Format:
+  show interfaces link-training status <interface_name>
+
+Arguments:
+  interface_name: [optional] Name of the interface to be shown. e.g: Ethernet0.
+                  If interface_name is not given, this command shows link training status for all interfaces.
+
+Example:
+  show interfaces link-training status
+  show interfaces link-training status Ethernet0
+
+Return:
+  error message if interface_name is invalid otherwise:
+
+admin@sonic:~$ show interfaces link-training status
+  Interface      LT Oper    LT Admin    Oper    Admin        Error Description
+-----------  -----------  ----------  ------  -------  -----------------------
+  Ethernet0      trained          on      up       up  unsupported transceiver
```

Comment:
I believe unsupported transceiver will not be relevant since we removed pmon related sections.

Reply:
HLD updated